### PR TITLE
User interface 2

### PR DESCRIPTION
**Pull request by julia57**
- need validation
- comments?

### CODE REVIEW

File: Assignment_6/ParkingUI.cpp [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+            continue;
+        }
+        std::vector<unsigned> new_arrangement;
+        GetUsersArrangement(NumberOfSlots, new_arrangement);
+        std::cout << "Rearranging process:" << std::endl;
+        parking.Rearrange(new_arrangement);
+        parking.PrintLastActions(std::cout);
+        std::cout << std::endl;
+    }
+    while (1) {
```
Comment (kris1319) [inline Assignment_6/ParkingUI.cpp]: Cool idea :) 
But stop... You'll recursively call methods `Run() -> Rearrengings() -> Run() -> Rearrangings() -> ....` and it's possible to get stackoverflow exception!

However you can notice that this code is very similar to the code in the `Run()` method. I think you can put this cycle into the `Run()` method and put request "Do you want to enter another arrangement (yes/no)?" at the end of the cycle-block, and reformulate this "Set a number of the parking places:" into "New parking is ready. Set a number of places:".

My idea is that `Run()` method should be the main method in this class but now you call it from another one (and also there is a bug in recursive callings).

File: Assignment_6/ParkingUI.h [hunk 2]
```diff
@@ -0,0 +1,11 @@
+#pragma once
+
+#include "arrangement.h"
+
+// ParkingUI class provides UI.
+class ParkingUI {
```
Comment (kris1319) [inline Assignment_6/ParkingUI.h]: The class have no fields so you can add `static` specifier to every method in this class and run it in the main function as `ParkingUI::Run();` :)

Comment (julia57) [inline Assignment_6/ParkingUI.h]: Done

File: Assignment_6/parking_ui.cpp [hunk 3]
```diff
@@ -0,0 +1,82 @@
+#include "parking_ui.h"
+
+void ParkingUI::GetUsersArrangement(int NumberOfSlots, std::vector<unsigned>& arrangement) {
```
Comment (kris1319) [inline Assignment_6/parking_ui.cpp]: Oh, I'm sorry I didn't notice it in my previous review. Can you rename `NumberOfSlots` to `number_of_slots` according to project's code style?

File: Assignment_6/parking_ui.cpp [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    std::cout << "Enter an arrangement of cars at the parking lot" << std::endl;
+    unsigned car;
+    for (int i = 0; i < NumberOfSlots; ++i) {
+        std::cin >> car;
+        arrangement.push_back(car);
+    }
+}
+
+void ParkingUI::Run() {
+    int NumberOfSlots;
```
Comment (kris1319) [inline Assignment_6/parking_ui.cpp]: Here also `NumberOfSlots` -> `number_of_slots `

File: Assignment_6/parking_ui.cpp [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    std::cout << "Set a number of the parking places:" << std::endl;
```
Comment (kris1319) [inline Assignment_6/parking_ui.cpp]: You don't need different code for the first running. You can use one infinite cycle.

File: Assignment_6/parking_ui.cpp [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+            std::cout << "Number of parking places can't be less than 0" << std::endl;
+        }
+    }
+}
+
+void ParkingUI::Rearrangings(int NumberOfSlots, CarParking& parking) {
+    if (NumberOfSlots == 0) {
+        return;
+    }
+    std::string user_answer;
```
Comment (kris1319) [inline Assignment_6/parking_ui.cpp]: Blank line before :)

**Comment by julia57** [comment]
About algorithm itself
1) About SetCarSlots method:  maybe it would be more convenient to concider an empty spot as anothr car and put it in a slot too. So we will be able to use car_slots[i] instead of cars_slots[i-1] in some places

2) I really think that we can avoid using used_cars. Here in Rearrange function we use iteration like
 for (auto car : arrangement_)
but
I think that maybe if we used index iteration: for (int i = 0; i < N; ++i) we could awoid extra memory. Because this way we don't need to see if the car is used. We just need to se if car is in the proper place or it is in unordered chain or cycle.

**Comment by agaga23** [comment]
According to @julia57 comment 1 about algorithm:
We can also have if(arrangment[i]) car_slots[arrangement[i]] = i; so we don't need to assume empty slot as car but we can use car_slots[i] instead of car_slots[i-1].

File: Assignment_6/parking_ui.cpp [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        arrangement.push_back(car);
+    }
+}
+
+void ParkingUI::Run() {
+    // variables
+    unsigned number_of_slots;
+    std::string user_answer;
+
+    while (1) {
```
Comment (agaga23) [inline Assignment_6/parking_ui.cpp]: I would change it into while (true) to have bool argument

File: Assignment_6/parking_ui.cpp [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+void ParkingUI::Rearrangings(int number_of_slots, CarParking& parking) {
+    if (number_of_slots == 0) {
+        return;
+    }
+
+    std::string user_answer;
+    while (1) {
+        std::cout << "Do you want to rearrange the cars (yes/no)?" << std::endl;
+        std::cin >> user_answer;
+        if (user_answer == "no") {
```
Comment (agaga23) [inline Assignment_6/parking_ui.cpp]: Maybe we could change checking user_answer as in Run()  (lines 34-40)

File: Assignment_6/arrangement.cpp [hunk 9]
```diff
@@ -29,9 +29,22 @@ std::vector<CarParking::Action> CarParking::GetLastActions() const {
...
+        std::swap(arrangement[action.from], arrangement[action.to]);
+        for (unsigned i : arrangement) {
+            out << i << " ";
+        }
+        std::cout << std::endl;
     }
+    std::cout << "Rearranged by " << step_counter << " steps" << std::endl;
 }
 
 bool CarParking::operator==(const CarParking& other) const {
```
Comment (agaga23) [inline Assignment_6/arrangement.cpp]: Do we use it somewhere? I think it costs us a lot to compare arrays.

File: Assignment_6/parking_ui.cpp (hunk 7 above)
Comment (julia57) [inline Assignment_6/parking_ui.cpp]: Is it better?

File: Assignment_6/arrangement.cpp [hunk 10]
```diff
@@ -29,9 +29,22 @@ std::vector<CarParking::Action> CarParking::GetLastActions() const {
...
+    std::vector<unsigned> arrangement = last_arrangement_;
+    for (unsigned i : arrangement) {
+        out << i << " ";
+    }
+    out << std::endl;
+    unsigned step_counter = 0;
     for (const auto& action : last_actions_) {
+        out << "Step " << ++step_counter << ":" << std::endl;
         out << "Move from " << action.from << " to " << action.to << std::endl;
+        std::swap(arrangement[action.from], arrangement[action.to]);
```
Comment (agaga23) [inline Assignment_6/arrangement.cpp]: In ImplementAction() Aida doesn't use swap. She writes the same thing using:
+    unsigned car = arrangement_[action.from];
+    arrangement_[action.to] = car;
+    arrangement_[action.from] = 0; 
I think we should have both functions similar. But I don't know which one should we change.

Comment (agaga23) [inline Assignment_6/arrangement.cpp]: In ImplementAction() Aida doesn't use swap. She writes the same thing using:
+    unsigned car = arrangement_[action.from];
+    arrangement_[action.to] = car;
+    arrangement_[action.from] = 0; 
I think we should have both functions similar. But I don't know which one should we change.

File: Assignment_6/parking_ui.cpp (hunk 7 above)
Comment (agaga23) [inline Assignment_6/parking_ui.cpp]: I've got a comment after assignment 2 from ssancho2: "Please use true/false for booleans." (And somewhere later in the code there is also while(1))

Comment (julia57) [inline Assignment_6/parking_ui.cpp]: okey:)

**Comment by julia57** [comment]
add pdf solution. Need a review

**Comment by agaga23** [comment]
It is very understandable and concise :) 
In paragraph 1: Your function must generate a serieS?? (S to delete?) (...) and the inputS (S to delete?)
In paragraph 2: You numbered places from 0 to N-1 so maybe in arrays there should be a0, ..., a(n-1) and b0, ..., b(n-1)
I would change 'fake' cycle to 'short' cycle
In paragraph "Goal": I would change the number of moves to (N - 1) - Q + K (this sequence is more logical for me)
In "proposition 1": ..less than by -> by less than

**Comment by julia57** [comment]
Thank you!
Agree, I'll change it now

**Comment by agaga23** [comment]
In proposition1 I would change the beginning a little bit for sth like this:
It is important hat by one step we can move only one car and as in the chain all cars must be moved at least once, we need to do at least t – 1 zero-transpositions.

I would change „0” into „zero”, as you use „zero” later

proposition2 : I don't understand the sentence „we can put a number in real cycle in the right place by 1 step”.Did you mean „in chain”?

Maybe proposition3 (fake-cycle): we don't need to move cars which are in fake-cycle. 
(Not to ommit this obvious element)

**Comment by julia57** [comment]
oh, it must be "we can't put a number in real cycle in the right place by 1 step"

**Comment by agaga23** [comment]
Ok, I undestand now :D

The last comment :) 

In MAIN THEOREM:
I would change as in previous comment "non-zero number" and "position -> final position":
As it was stated: each of them needs at least one step to go into final position

I would change the sequence to N – 1 – Q + K (at the end and in Main Theorem)

**Comment by agaga23** [comment]
I have read it again and I found my typo, sorry

In proposition1: It is important hat -> that

**Comment by Laucalle** [comment]
Please add this to the Makefile:

parking_ui.o : parking_ui.cpp $(GTEST_HEADERS)
	$(CXX) $(CPPFLAGS) $(CXXFLAGS) -c parking_ui.cpp

**Comment by julia57** [comment]
Ok! Thank you!

**Pull request merged**